### PR TITLE
[AMS-1122] Redis connection loss handling

### PR DESCRIPTION
**Pull request by denisivan0v**

### CODE REVIEW

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+using StackExchange.Redis;
+
+namespace NuClear.VStore.Locks
+{
+    public sealed class ReliableRedLockFactory : IDistributedLockFactory, IDisposable
+    {
+        private const int DefaultConnectionTimeout = 1000;
+        private const int DefaultSyncTimeout = 1000;
+        private const int DefaultKeepAlive = 1;
```
Comment (xakep139) [inline src/VStore/Locks/ReliableRedLockFactory.cs]: Move these values in config?

File: tests/VStore.UnitTests/ImageRendering/MemoryConsumptionTests.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ public MemoryConsumptionTests(ITestOutputHelper output)
             Configuration.Default.MemoryManager = ArrayPoolMemoryManagerFactory.CreateWithLimitedPooling();
         }
 
-        [Fact]
+        [Fact(Skip = "Disabled")]
```
Comment (xakep139) [inline tests/VStore.UnitTests/ImageRendering/MemoryConsumptionTests.cs]: Do we plan to return these tests in future?

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        {
+            switch (endPoint)
+            {
+                case DnsEndPoint dnsEndPoint:
+                    return $"{dnsEndPoint.Host}:{dnsEndPoint.Port}";
+                case IPEndPoint ipEndPoint:
+                    return $"{ipEndPoint.Address}:{ipEndPoint.Port}";
+            }
+
+            return endPoint.ToString();
```
Comment (xakep139) [inline src/VStore/Locks/ReliableRedLockFactory.cs]: Move to `default` clause?

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        private static void OnInternalError(object sender, InternalErrorEventArgs args)
+            => _redisLogger.LogWarning(
+                args.Exception,
+                "InternalError: {endpoint} ConnectionType: {connectionType} Origin: {origin}",
+                GetFriendlyName(args.EndPoint),
+                args.ConnectionType,
+                args.Origin);
+
+        private static void OnInternalError(object sender, RedisErrorEventArgs args)
+            => _redisLogger.LogWarning("ErrorMessage: {endpoint} Message: {message}", GetFriendlyName(args.EndPoint), args.Message);
```
Comment (xakep139) [inline src/VStore/Locks/ReliableRedLockFactory.cs]: `ErrorMessage` -> `Error`?

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                                try
+                                {
+                                    logger.LogTrace("Checking RedLock endpoint {endpoint} for availablity.", GetFriendlyName(endpoint));
+                                    var server = multiplexer.GetServer(endpoint);
+                                    server.Ping();
+                                    logger.LogTrace("RedLock endpoint {endpoint} is available.", GetFriendlyName(endpoint));
+                                }
+                                catch
+                                {
+                                    logger.LogWarning(
```
[the remark this answers is not in the document]
Comment (denisivan0v) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: Agree, but we don't want to log stack trace here

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                                        "RedLock endpoint {endpoint} is unavailable. All connections will be recreated.",
+                                        GetFriendlyName(endpoint));
+                                    recreationNeeded = true;
```
[the remark this answers is not in the document]
Comment (denisivan0v) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: Yes, it's better to have more information about connectivity loss

File: src/VStore/Locks/ReliableRedLockFactory.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (denisivan0v) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: Done in c7c1338

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+        private static void OnInternalError(object sender, InternalErrorEventArgs args)
+            => _redisLogger.LogWarning(
+                args.Exception,
+                "InternalError: {endpoint} ConnectionType: {connectionType} Origin: {origin}",
+                GetFriendlyName(args.EndPoint),
+                args.ConnectionType,
+                args.Origin);
+
+        private static void OnInternalError(object sender, RedisErrorEventArgs args)
```
[the remark this answers is not in the document]
Comment (denisivan0v) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: Done in c7c1338

File: src/VStore/Locks/ReliableRedLockFactory.cs (hunk 3 above)
Comment (denisivan0v) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: Done in c7c1338

File: tests/VStore.UnitTests/ImageRendering/MemoryConsumptionTests.cs (hunk 2 above)
Comment (denisivan0v) [reply tests/VStore.UnitTests/ImageRendering/MemoryConsumptionTests.cs]: The answer is here https://github.com/2gis/nuclear-vstore/pull/163#issuecomment-389404376

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+                                CommandMap = CommandMap.Create(new HashSet<string> { "SUBSCRIBE" }, available: false),
+                                Password = lockOptions.Password,
+                                ConnectTimeout = lockOptions.ConnectionTimeout ?? DefaultConnectionTimeout,
+                                SyncTimeout = lockOptions.SyncTimeout ?? DefaultSyncTimeout,
+                                KeepAlive = keepAlive,
+                                // Time (seconds) to check configuration. This serves as a keep-alive for interactive sockets, if it is supported.
+                                ConfigCheckSeconds = keepAlive
+                            };
+
+                        var multiplexer = await ConnectionMultiplexer.ConnectAsync(redisConfig, logWriter);
```
Comment (xakep139) [inline src/VStore/Locks/ReliableRedLockFactory.cs]: Why do we create separate multiplexer on each endpoint instead of creating one multiplexer with multiple endpoints?

Comment (denisivan0v) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: The Redlock algorithm operates with many Redis instances, so each we need to connect every instance independently, multiplexer = Redis instance (we have only 1 node clusters contains just master node).

File: src/VStore/Locks/ReliableRedLockFactory.cs [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+using RedLockNet.SERedis;
+using RedLockNet.SERedis.Configuration;
+
+using StackExchange.Redis;
+
+namespace NuClear.VStore.Locks
+{
+    public sealed class ReliableRedLockFactory : IDistributedLockFactory, IDisposable
+    {
+        private const int DefaultConnectionTimeout = 1000;
```
Comment (denisivan0v) [inline src/VStore/Locks/ReliableRedLockFactory.cs]: @xakep139 there was a comment about this defaults (maybe you delete it, maybe it just vanished), you proposed to move them to config. 
The point is, everything configured via config (https://github.com/2gis/nuclear-vstore/pull/162/files#diff-20eabfdca502c44d5509088e3e1d021dR106), but if not, these values will be applied.

Comment (xakep139) [reply src/VStore/Locks/ReliableRedLockFactory.cs]: Yeah, I've found out this before deleting this comment ))

**Comment by denisivan0v** [comment]
@OlgilvieMaurice thanks for this test case on boundary conditions! It is really makes sense for a straightforward implementation of connection recreating. 
However, as you can see here 

https://github.com/2gis/nuclear-vstore/blob/fix/redis-reconnection/src/VStore/Locks/ReliableRedLockFactory.cs#L192 

we creating new Redis connections in advance (incapsulated by  `_innerFactory `) and than replacing them current connections with the new ones atomically.

So, if connections recreation process was successful (for example, in case of transient network failures) than we always have as many active connections as possible and there is no break of fault tolerance. 
In case if we can't recreate connections, we essentially have non-transient network problems and it's highly possible that other parts of application are down too.

**Comment by OlgilvieMaurice** [comment]
@denisivan0v What if the reason a Redis instance goes down isn't a network failure, but some normal, or at least not unexpected, behavior, e.g. in k8s, a pod gets evicted?

**Comment by denisivan0v** [comment]
@OlgilvieMaurice it is still the same from the client point of view.
There was some race condition situations that I've solved (or at least found a tradeoff) in recent commits.

**Comment by OlgilvieMaurice** [comment]
lgtm
